src/graphapi/enginefinder.cpp

```diff
@@ -0,0 +1,167 @@
+        {
+            mHandle = *static_cast<miopenHandle_t*>(arrayOfElements);
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+        break;
+
+    case MIOPEN_ATTR_OPERATIONGRAPH_OPS:
+        if(attributeType == MIOPEN_TYPE_BACKEND_DESCRIPTOR && elementCount > 0)
+        {
+            std::vector<miopenBackendDescriptor_t> ops;
+            ops.reserve(elementCount);
+
+            OpGraphBuilder builder;
+
+            std::for_each_n(static_cast<miopenBackendDescriptor_t*>(arrayOfElements),
+                            elementCount,
+                            [&ops, &builder](miopenBackendDescriptor_t apiDescriptor) {
+                                BackendDescriptor& backendDescriptor = deref(apiDescriptor);
+                                if(backendDescriptor.isFinalized())
+                                {
+                                    OpNode* node = backendDescriptor.getOperation();
+                                    if(node != nullptr && !builder.hasNode(node))
+                                    {
+                                        builder.addNode(node);
+                                        ops.push_back(apiDescriptor);
+                                        return;
+                                    }
+                                }
+                                MIOPEN_THROW(miopenStatusBadParm);
+                            });
+
+            mOps            = std::move(ops);
+            mOpGraphBuilder = std::move(builder);
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+        break;
+
+    default: MIOPEN_THROW(miopenStatusBadParm);
+    }
+}
+
+void miopen::graphapi::BackendOperationGraphDescriptor::finalize()
+{
+    if(mFinalized)
+    {
+        MIOPEN_THROW(miopenStatusNotInitialized);
+    }
+
+    if(mHandle == nullptr || mOps.empty())
+    {
+        MIOPEN_THROW(miopenStatusBadParm);
+    }
+
+    mOpGraph = std::move(mOpGraphBuilder).build();
+
+    // TODO: Find solutions for the operation graph
+
+    mFinalized = true;
+}
+
+void miopen::graphapi::BackendOperationGraphDescriptor::getAttribute(
+    miopenBackendAttributeName_t attributeName,
+    miopenBackendAttributeType_t attributeType,
+    int64_t requestedElementCount,
+    int64_t* elementCount,
+    void* arrayOfElements)
+{
+    if(!mFinalized)
+    {
+        MIOPEN_THROW(miopenStatusNotInitialized);
+    }
+
+    switch(attributeName)
+    {
+    case MIOPEN_ATTR_OPERATIONGRAPH_HANDLE:
+        if(attributeType == MIOPEN_TYPE_HANDLE && requestedElementCount == 1)
+        {
+            *elementCount                                  = 1;
+            *static_cast<miopenHandle_t*>(arrayOfElements) = mHandle;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusBadParm);
+        }
+        break;
+
+    case MIOPEN_ATTR_OPERATIONGRAPH_OPS:
```

Comment:
The `getAttribute` here is also strange in the sense that it's only useful if the user specifies the count as the number of nodes in the graph. If it's lesser, we don't know which nodes to return. I think we shouldn't support `getAttribute` for this case but I'm not insisting. I think the case where count == num nodes is useful but then there is not way to get the count of nodes from a graph in the C API.

Reply:
We should return descriptors the same as user has set them but not going beyond the provided storage cutting off the tail. The number of descriptors we actually have should be returned in *elementCount. A user can correct the storage size by calling getAttribute with requestedCount=0 and receiving the actual number in *elementCount; then the user calls it again with the right requestedCount value.